# Determinar modelo

Despues de estudiar cada modelo individualmente con el dataset con configuraciones iniciales estandar se determino que la mejor oopcion a explorar a fondo es el Gradient Boosting Classifier

In [20]:
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import GradientBoostingClassifier, RandomForestClassifier
from sklearn.metrics import accuracy_score

In [21]:
train = pd.read_csv('toAnalyze.csv')
test = pd.read_csv('test.csv')

# Separar las características (X) y la variable objetivo (y)
x = train.drop(columns=['smoking'])  # Todas las columnas excepto 'smoking'
y = train['smoking']  # La columna 'smoking'

X_train, X_val, y_train, y_val = train_test_split(x, y, test_size=0.2, random_state=42)

In [ ]:
train = pd.read_csv('toAnalyze.csv')
test = pd.read_csv('test.csv')

# Separar las características (X) y la variable objetivo (y)
x = train.drop(columns=['smoking'])  # Todas las columnas excepto 'smoking'
y = train['smoking']  # La columna 'smoking'

X_train, X_val, y_train, y_val = train_test_split(x, y, test_size=0.2, random_state=42)

In [22]:
# Escalar los datos
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_val_scaled = scaler.transform(X_val)

### Lineal Regression

In [23]:
lr_model = LogisticRegression(max_iter = 1000, random_state = 42)
lr_model.fit(X_train_scaled, y_train)

y_val_pred = lr_model.predict(X_val_scaled)
accuracy_lr = accuracy_score(y_val, y_val_pred)

### SVM

In [24]:
svm_model = SVC(kernel = 'rbf', C = 1,random_state = 42)
svm_model.fit(X_train_scaled, y_train)

y_val_pred = svm_model.predict(X_val_scaled)
accuracy_svm = accuracy_score(y_val, y_val_pred)

### Neural Network

In [25]:
nn_model = MLPClassifier(hidden_layer_sizes = (100, ), activation = 'relu', max_iter = 1000, random_state = 42)
nn_model.fit(X_train_scaled, y_train)

y_val_pred = nn_model.predict(X_val_scaled)
accuracy_nn = accuracy_score(y_val, y_val_pred)

### GBM

In [26]:
gbm_model = GradientBoostingClassifier(n_estimators = 100, learning_rate = 0.1, max_depth = 3, random_state = 42)
gbm_model.fit(X_train_scaled, y_train)

y_val_pred = gbm_model.predict(X_val_scaled)
accuracy_gbm = accuracy_score(y_val, y_val_pred)

### Random Forest

In [27]:
rf_model = RandomForestClassifier(n_estimators = 100, max_depth = None, random_state = 42)
rf_model.fit(X_train, y_train)

y_val_pred = rf_model.predict(X_val_scaled)
accuracy_rf = accuracy_score(y_val, y_val_pred)

d:\Users\Lenovo\Anaconda3\envs\eegGPU\lib\site-packages\sklearn\base.py:464: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


### Resumen

In [28]:
print(f"Accuracy of Logistic Regression: {accuracy_lr}")
print(f"Accuracy of SVM: {accuracy_svm}")
print(f"Accuracy of Neural Network: {accuracy_nn}")
print(f"Accuracy of Gradient Boosting Machine: {accuracy_gbm}")
print(f"Accuracy of Random Forest: {accuracy_rf}")

Accuracy of Logistic Regression: 0.7464209468793168
Accuracy of SVM: 0.765509230189627
Accuracy of Neural Network: 0.7650383021474318
Accuracy of Gradient Boosting Machine: 0.7690882833103102
Accuracy of Random Forest: 0.5766356900665578


# Busqueda de Hiperparametros

In [29]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split, RandomizedSearchCV
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

In [30]:
train = pd.read_csv('toAnalyze.csv')
test = pd.read_csv('test.csv')

# Separar las características (X) y la variable objetivo (y)
x = train.drop(columns=['smoking'])  # Todas las columnas excepto 'smoking'
y = train['smoking']  # La columna 'smoking'

X_train, X_val, y_train, y_val = train_test_split(x, y, test_size=0.2, random_state=42)

In [31]:
gbm_model = GradientBoostingClassifier(random_state = 42)

param_distributions = {
    'n_estimators' : [50, 100, 200, 300],
    'learning_rate' : [0.01, 0.05, 0.1, 0.2],
    'max_depth' : [3, 4, 5, 6],
    'min_samples_split' : [2, 5, 10],
    'min_samples_leaf' : [1, 2, 4],
    'subsample' : [0.7, 0.8, 0.9, 1]
}

In [32]:
random_search = RandomizedSearchCV(estimator = gbm_model, param_distributions = param_distributions, n_iter = 100, scoring = 'accuracy', n_jobs = -1, cv = 3, verbose = 2, random_state = 42)

random_search.fit(X_train, y_train)

print("Best hyperparameters: ", random_search.best_params_)
print("Best accuracy: ", random_search.best_score_)

Fitting 3 folds for each of 100 candidates, totalling 300 fits
Best hyperparameters:  {'subsample': 0.9, 'n_estimators': 300, 'min_samples_split': 5, 'min_samples_leaf': 4, 'max_depth': 6, 'learning_rate': 0.1}
Best accuracy:  0.7780681925214279


### Resultados:

- Best hyperparameters:  {'subsample': 0.9, 'n_estimators': 300, 'min_samples_split': 5, 'min_samples_leaf': 4, 'max_depth': 6, 'learning_rate': 0.1}
- Best accuracy:  0.7780681925214279